### ISSUE TITLE
Warn about security pitfalls of `find -exec` in favor of -execdir

### ISSUE THREAD

**Issue by rpdelaney** [koalaman/shellcheck#1786]
#### For new checks and feature suggestions
- [x] shellcheck.net (i.e. the latest commit) currently gives no useful warnings about this
- [x] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/usr/bin/env sh
find /tmp -path /tmp/umsp/passwd -exec /bin/rm \;
```

#### Here's what shellcheck currently says:
```
No issues detected!
```

#### Here's what I wanted or expected to see:
```
Line 2:
find /tmp -path /tmp/umsp/passwd -exec /bin/rm
                                 ^-- SCXXXX: -exec is insecure. Use -execdir.

```

### Rationale
shellcheck already warns about [problems with -exec and filenames](https://www.gnu.org/software/findutils/manual/html_node/find_html/Problems-with-_002dexec-and-filenames.html#Problems-with-_002dexec-and-filenames) in [SC2156](https://github.com/koalaman/shellcheck/wiki/SC2156). 

But there is another problem with `-exec`. The [manpage for GNU find](http://man7.org/linux/man-pages/man1/find.1.html) has cryptic warnings:

> -exec command ;
>...There are unavoidable security problems surrounding use of the -exec action; you should use the -execdir option instead.

>-execdir command ;
>-execdir command {} +
>
>...This a much more secure method for invoking commands, as it avoids race conditions during resolution of the paths to the matched files...

Later:

> There are security problems inherent in the behaviour that the POSIX standard specifies for find, which therefore cannot be fixed.  For example, the -exec action is inherently insecure, and -execdir should be used instead. 

The GNU [Finding Files manual](https://www.gnu.org/software/findutils/manual/html_node/find_html/Security-Considerations-for-find.html#Security-Considerations-for-find) has a full explanation.

Arguably, `-execdir` is not serviceable in all contexts, and `shellcheck` cannot be expected to know when the user's threat model includes these kinds of timing attacks by hostile users on the system. But we often warn about such problems anyway, such as in SC2156, and allow the user to disable the warning if it is not relevant.

**Comment by lordfolken** [comment]
There is another reason to advise against -exec:
 Find aborts on first failure of an -exec command, but it itself exits with return code 0. 
Small example:
```bash
touch 1 2 3
find
.
./1
./2
./3
find ./ -exec cat "{}" \;
cat: ./: Is a directory
echo $?
0

**Comment by mcandre** [comment]
Even aside from security find's `-exec` is fragile in terms of basic safety. It is not guaranteed to propagate the presence of a subprocess's failing exit code back to the overall `find` exit code.

Any commands performed on `find` results should be accomplished with either `xargs` or a `for` loop. Unknown how xargs or for loops handle ASCII spaces by default among POSIX sh and the various sh implementations.

What we do know, is that the POSIX specification maintainers absurdly refuse to adopt the `-print0` flag as a recognized, portable standard. When the `-print0` flag is present, the script should explicitly call `gfind`, instead of implicitly assuming portable POSIX `find`. When absent, then only some arcane `read` command with an appropriate `IFS` shadow to correctly iterate over the find results. Don't let the `IFS` declaration spill into successive commands. Don't let executable shell scripts, or shell scripts absent file extensions likely to be run as `sh <path>`, neglect to `unset IFS` immediately after the shebang line. If UNIX find is a gas station, then the -exec flag is an adjacent fireworks store. Preferably avoid both, but at the very least avoid the second one.
